doc/source/developer_overview.rst

```diff
@@ -156,50 +156,53 @@ Because ``ament_tools`` and ``ament_package`` are ament Python packages, they ca
+The ``rcl`` |API| is a slightly higher level |API| which is used to implement the |client libraries| and does not touch the middleware implementation directly.
 
-In addition to these two static |API's| there is the ``rosidl`` |API|, which is more complicated because it usually generated for each message and because it can be exposed directly to users.
-The ``rosidl`` |API| simply defines what code is generated for messages in each language and the ``rosidl`` implementation contains code for converting from ROS message types to and from native types for the middleware implementation (again in this case vendor specific DDS types).
-The ``rosidl`` C |API| can be reused by other languages or not, depending on the scenario.
+In addition to these two static |APIs| there is the ``rosidl`` |API|, which is more complicated because it is usually generated for each message and because it can be exposed directly to users.
+The ``rosidl`` *interface* simply defines what code is to be generated for messages in each programming language that has a |client library|, while the ``rosidl`` *implementation* contains the code for generating that code, i.e. converting ROS message types to and from native types for the middleware implementation (vendor specific DDS types).
+The ``rosidl`` C implementation can be reused by other languages that can interface with C, e.g. Python.
 
 .. figure:: images/ros_client_library_api_stack.png
    :alt: ros2 software stack
 
-As the diagram shows, these |API's| are stacked such that the ROS developer uses the client library |API|, e.g. ``rclcpp``, to write the implementation of their code (executable or library).
-The implementation of the client library, e.g. ``rclcpp``, then uses both the ``rcl`` and the ``rosidl`` interfaces.
+As the diagram shows, these |APIs| are stacked such that the ROS developer uses the client library |API|, e.g. ``rclcpp``, to write the implementation of their code (executable or library).
+The implementation of the client library, e.g. ``rclcpp``, then uses both the ``rcl`` and the ``rosidl`` interfaces, for providing access to the ROS graph and for communicating with other nodes through standardized ROS messages, respectively.
 The ``rcl`` implementation is provided by the ``rcl`` |package| and uses the ``rmw`` |API| to implement itself, leaving parts unimplemented where message type specific code is required.
-Then the ``rosidl`` implementation implements symbols from both the generated headers and some type specific functions in the ``rcl`` and ``rmw`` interfaces.
-The actual library which contains this implementation is provided by |packages| with messages in them, e.g. ``geometry_msgs``, but the code is indirectly provided by a language specific ``rosidl`` |package|, e.g. ``rosidl_generator_cpp``.
+Then the ``rosidl`` implementation implements symbols from both the generated headers and some type specific functions in the ``rcl`` and ``rmw`` interfaces that were left unimplemented.
```

Comment:
(the generated headers, type specific functions used by rcl and rmw)

The generated headers are the message specific headers that contain the C/C++ structs for each message type.

The "unimplemented" type specific functions used by rmw and rcl are the "get type support for X" functions, where X is a message type. However, I think those functions used to be under the `rmw` namespace, but are now in the `rosidl` namespace. I'll have to read this section again to see if it needs to be updated more.